From deb60e288937839be5cc6a240d193a73163fabcc Mon Sep 17 00:00:00 2001
From: Karthikeyan Kathirvel <quic_kathirve@quicinc.com>
Date: Fri, 27 May 2022 17:45:19 +0530
Subject: [PATCH] sigma-dut: set su beamformee based on hw vht_capab

Set SU beamformee based on the hw vht capabilities, for IPQ6018 SU
beamformee in VHT mode doesn't supported so this has to be set based
on hw capabilities

Signed-off-by: Karthikeyan Kathirvel <quic_kathirve@quicinc.com>
---
 ap.c        | 13 ++++++++++---
 sigma_dut.h |  3 +++
 2 files changed, 13 insertions(+), 3 deletions(-)

--- a/ap.c
+++ b/ap.c
@@ -9551,12 +9551,14 @@ skip_key_mgmt:
 		     dut->ap_tx_stbc == VALUE_ENABLED || dut->ap_mu_txBF ||
 		     dut->ap_ampdu_exp || dut->ap_max_mpdu_len ||
 		     dut->ap_chwidth == AP_160 ||
-		     dut->ap_chwidth == AP_80_80) &&
+		     dut->ap_chwidth == AP_80_80 ||
+		     dut->ap_su_txBF) &&
 		     check_band(dut, BAND_5G, conf_counter)) {
-			fprintf(f, "vht_capab=%s%s%s%s%s%s",
+			fprintf(f, "vht_capab=%s%s%s%s%s%s%s",
 				dut->ap_sgi80 ? "[SHORT-GI-80]" : "",
 				dut->ap_txBF ?
-				"[SU-BEAMFORMER][SU-BEAMFORMEE][BF-ANTENNA-2][SOUNDING-DIMENSION-2]" : "",
+				"[SU-BEAMFORMER][BF-ANTENNA-2][SOUNDING-DIMENSION-2]" : "",
+				dut->ap_su_txBF ? "[SU-BEAMFORMEE]":"",
 				(dut->ap_ldpc == VALUE_ENABLED) ?
 				"[RXLDPC]" : "",
 				(dut->ap_tx_stbc == VALUE_ENABLED) ?
@@ -10511,6 +10513,10 @@ static int wiphy_info_handler(struct nl_
 	    !(dut->hw_modes.ap_he_phy_capab[33 / 8] & (1 << (33 % 8))))
 		dut->ap_mu_txBF = 0;
 
+	if ((dut->program == PROGRAM_VHT || dut->program == PROGRAM_HE) &&
+	    !(dut->hw_modes.vht_capab & IEEE80211_VHT_CAP_SU_BEAMFORMEE_CAPABLE))
+		dut->ap_su_txBF = 0;
+
 	return 0;
 }
 
@@ -10672,6 +10678,7 @@ static enum sigma_cmd_result cmd_ap_rese
 	dut->ap_rx_amsdu = VALUE_NOT_SET;
 	dut->ap_txBF = 0;
 	dut->ap_mu_txBF = 0;
+	dut->ap_su_txBF = 1;
 	dut->ap_chwidth = AP_AUTO;
 	dut->ap_ampdu_exp = 0;
 	dut->ap_max_mpdu_len = 0;
--- a/sigma_dut.h
+++ b/sigma_dut.h
@@ -101,6 +101,8 @@ struct sigma_dut;
 /* Set default operating channel width 80 MHz */
 #define VHT_DEFAULT_OPER_CHWIDTH AP_80_VHT_OPER_CHWIDTH
 
+#define IEEE80211_VHT_CAP_SU_BEAMFORMEE_CAPABLE 0x00001000
+
 typedef unsigned int u32;
 typedef uint16_t u16;
 typedef unsigned char u8;
@@ -721,6 +723,7 @@ struct sigma_dut {
 	} ap_vht_chwidth;
 	int ap_txBF;
 	int ap_mu_txBF;
+	int ap_su_txBF;
 	enum ap_regulatory_mode {
 		AP_80211D_MODE_DISABLED,
 		AP_80211D_MODE_ENABLED,
